base/utf.jl

```diff
@@ -0,0 +1,993 @@
+                # next character *must* be a trailing surrogate character
+                ch = dat[pos += 1]
+                !is_surrogate_trail(ch) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos, ch)
+                num4byte += 1
+                (options & UTF_NO_SURROGATES) != 0 && utf_errfunc(UTF_ERR_SURROGATE, pos, ch)
+                flags |= UTF_SURROGATE
+            else
+                utf_errfunc(UTF_ERR_NOT_LEAD, pos, ch)
+            end
+        end
+    end
+    totalchar, flags | (num3byte == 0 ? 0 : UTF_UNICODE3) | (num4byte == 0 ? 0 : UTF_UNICODE4), num4byte, num3byte, num2byte
+end
+
+function check_string_abs(str::AbstractString, options::Integer=0)
+    local ch::UInt32
+    local totalchar=0, num2byte=0, num3byte=0, num4byte=0, flags::UInt=0
+    local pos = start(str)
+    local len = endof(str)
+    @inbounds while pos < len
+        ch, pos = next(str, pos)
+        totalchar += 1
+        if ch > 0x7f
+            if ch < 0x100
+                num2byte += 1
+                flags |= UTF_LATIN1
+            elseif ch < 0x800
+                num2byte += 1
+                flags |= UTF_UNICODE2
+            elseif ch > 0xffff
+                (ch > 0x10ffff) && utf_errfunc(UTF_ERR_INVALID, pos, ch)
+                num4byte += 1
+            elseif !is_surrogate_char(ch)
+                num3byte += 1
+            elseif is_surrogate_lead(ch)
+                pos == len && utf_errfunc(UTF_ERR_MISSING_SURROGATE, pos, ch)
+                # next character *must* be a trailing surrogate character
+                ch, pos = next(str, pos)
+                !is_surrogate_trail(ch) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos, ch)
+                num4byte += 1
+                (options & UTF_NO_SURROGATES) != 0 && utf_errfunc(UTF_ERR_SURROGATE, pos, ch)
+                flags |= UTF_SURROGATE
+            else
+                utf_errfunc(UTF_ERR_NOT_LEAD, pos, ch)
+            end
+        end
+    end
+    totalchar, flags | (num3byte == 0 ? 0 : UTF_UNICODE3) | (num4byte == 0 ? 0 : UTF_UNICODE4), num4byte, num3byte, num2byte
+end
+
+# Quickly copy and set trailing \0
+macro return_fast_utf_copy(T1, T2, len, dat)
+    quote
+        @inbounds return $(esc(T1))(setindex!(copy!(Vector{$(esc(T2))}($(esc(len))), $(esc(dat))), 0, $(esc(len))))
+    end
+end
+
+# Get rest of character ch from 2-byte UTF-8 sequence in str, update pos and return character
+macro get_utf8_2!(str, pos, ch)
+    quote
+        (($(esc(ch)) & 0x1f) << 6) | ($(esc(str))[$(esc(pos)) += 1] & 0x3f)
+    end
+end
+
+# Get rest of character ch from 3-byte UTF-8 sequence in str, update pos and return character
+macro get_utf8_3!(str, pos, ch)
+    quote
+        ($(esc(pos)) += 2 ;
+         (($(esc(ch)) & 0xf) << 12)
+          | (UInt32($(esc(str))[$(esc(pos))-1] & 0x3f) << 6)
+          | ($(esc(str))[$(esc(pos))] & 0x3f))
+    end
+end
+
+# Get rest of character ch from 3-byte UTF-8 sequence in str, update pos and return character
```

Comment:
Just like I commented before about `@get_continuation`, is there any reason why these should be macros instead of functions?